### ISSUE TITLE
Aggregate using diff_first or diff_last will leave a gap at first entry

### ISSUE THREAD

**Issue by crnjan** [openhab/openhab-webui#912]
## The problem

Having a HP that logs usage time - incrementally, number of hours. Used this values to display number of hours HP was producing heat for either DHW or underfloor heating:

```
   ...
    - component: oh-aggregate-series
      config:
        name: DHW
        gridIndex: 0
        xAxisIndex: 0
        yAxisIndex: 0
        type: bar
        item: gBasement_BoilerRoom_HP_DHW_InOperation_HP
        dimension1: date
        aggregationFunction: diff_first
        service: influxdb
  ...
```

result:

<img width="893" alt="Screenshot 2021-02-16 at 21 16 37" src="https://user-images.githubusercontent.com/3818211/108119978-dbd24180-70a0-11eb-9991-2bf5ef3c7e36.png">

All good, just first bar is missing. Looking at the code, this is expected.

## Your suggestion

We could extend the period and fetch one item "before" first period and use that one for reference to calculate first diff - but might be problematic to define what time interval to use to fetch that item?

I used a bit different approach and introduced a new aggregator `diff` that just takes first and last value from within the group itself.

```
    case 'diff':
      aggregate = arr[1][arr[1].length - 1] - arr[1][0]
```

result:

<img width="872" alt="Screenshot 2021-02-16 at 21 23 08" src="https://user-images.githubusercontent.com/3818211/108120577-c873a600-70a1-11eb-9799-2543a8d5233a.png">

I look forward to any feedback or comments - and if `diff` aggregator makes sense to others as well, I can prepare a PR ...

**Comment by ghys** [comment]
I'm not against a "diff" aggregation, but it could mislead people into believing the interval between 2 values is the exact aggregation period selected when it's not, i.e. with a hourly period if would be e.g. 59 minutes if you have data points every minute, 55 minutes if you have data points every 5 minutes etc.
The Persistence API has a `boundary` option:
_("Gets one value before and after the requested period.)_
I think I ran into problems when using it. Might be worth another look though - perhaps with a mechanism to only including it when querying data for aggregate series (not time series) with a diff_* aggregator.

**Comment by crnjan** [comment]
Did a quick spike and seems as `boundary` does return additional value, thx for the hint. Will try to use that value for diff_* aggregator when I find time to do so ..

**Comment by rumexcrisp** [comment]
is there any news on this issue? i would be very interested in a solution to this problem

**Comment by seimenb** [comment]
@crnjan Hi, could you finaly find a solution for that issue? I ran into the same problem, not seeing the first bar.

**Comment by cweitkamp** [comment]
The linked issue in openHAB core has been fixed some time ago (see https://github.com/openhab/openhab-core/pull/2668). It fixes an problem when `boundary` parameter is set to `true`.

Is anyone still working on this aggregated chart issue?

**Comment by crnjan** [comment]
Did a quick spike in the past (exactly one year since I opened issue mentioned above -  https://github.com/openhab/openhab-core/issues/2665 :) ) and what I was "afraid" is the fact that different users have different things reporting different values at different intervals with different precision ... i.e. last value from previous interval might not be exactly at the point (in time) where you want it (and pretty much it never is) so one cannot calculate exact diff - so users might report - I produced exactly XXXX of that and that and OH is reporting YYYY ... to be honest did not want to go down that rabbit hole ... but might take another look when I find time to do so ... but as said, there are limitations that might not be super obvious ...

**Comment by crnjan** [comment]
Played with aggregator(s) and remembered the main issue - `diff_first`, but let's first start with my POC - added (hardcoded for testing) the `boundary=true` parameter when requesting data, assuming (and please correct me if I'm wrong, maybe @cweitkamp) the first returned value (altho it's timestamp is set to `starttime` when returned), it's actually first persisted datapoint *before* `starttime` - using `diff_last` (taking difference between *last* datapoint within each interval), we can than take first returned datapoint (with combination of  `boundary=true`) when calculating difference for first interval, since it's actually the last datapoint of previous interval ... so long story short, under above assumptions, we can update

```
case 'diff_last':
      aggregate = idx < 1 ? NaN : arr[1][arr[1].length - 1] - values[idx - 1][1][values[idx - 1][1].length - 1]
```

to

```
case 'diff_last':
      if (idx < 1) {
        aggregate = arr[1][arr[1].length - 1] - arr[1][0]
      } else {
        aggregate = arr[1][arr[1].length - 1] - values[idx - 1][1][values[idx - 1][1].length - 1]
      }
```

which renders
<img width="830" alt="Screenshot 2023-01-04 at 21 18 02" src="https://user-images.githubusercontent.com/3818211/210645063-76afd0f0-5219-4fce-91cf-f51a2516a3cc.png">

compared to what we get without the change

<img width="1283" alt="Screenshot 2023-01-04 at 21 36 43" src="https://user-images.githubusercontent.com/3818211/210645264-8df87aa2-47c5-4c64-8aaf-ed3a9761e190.png">

So if my understanding is correct, `diff_last` could be solved using `boundary=true`. But `diff_first` is not - it uses first point of each interval to calculate the diff, so we would need to get first point of interval before `starttime` (in above chart first point of November since chart is for December) and that is not solvable using simply boundary=true` but we would need to tweak query (`starttime`) so we ensure we include that in response, imho ...

I'll try if I make it work for `diff_last` beyond POC ...

**Comment by cweitkamp** [comment]
@crnjan Thanks for creating the POC. I will check tomorrow which values are returned when adding `boundary=true` to the query.

wrt `diff_first`: I have to admit that I do not understand the implemented logic for it. I have a weekly aggregation for my power consumption. When using `diff_first` all aggregated values are shifted by one day. Value of Monday is shown on Tuesday, value of Tuesday is shown on Wednesday and so on.

![Screenshot_20230104-221511~2](https://user-images.githubusercontent.com/5131747/210652244-fa201b71-ff38-4f96-89c2-e971e1be84d3.jpg)

Whereas when using `diff_last` the first day (Monday) is missing,, the rest is correct.

![Screenshot_20230104-221532~2](https://user-images.githubusercontent.com/5131747/210652412-3b50459e-4001-4d18-b670-59ebb7bdb970.jpg)

As comparison the aggregated weekly chart by volkszaehler.org Web-App - the tool I am using to get my meter readings via IR sensor and SML protocol.

![Screenshot_20230104-221546~2](https://user-images.githubusercontent.com/5131747/210652660-5650516c-eaac-4638-8797-a2fd20ec9df1.jpg)

**Comment by crnjan** [comment]
First data is fetched for given interval, in my case above for month. Then all datapoints are grouped by day, so we get array of days, each containing array of datapoints within each day, something like:

```
[[/* datapoints of day 1*/], [/* datapoints of day 2*/], ...]
```

The difference between `diff_first` and `diff_last` is - `diff_first` takes first point of current array (i.e. day 2) and subtracts first point of previous day, in above case, day 1; while `diff_last` takes last ...

**Comment by crnjan** [comment]
Created a test jar, please run

```
bundle:update org.openhab.ui https://github.com/crnjan/openhab-webui/raw/feature/diff_poc/bundles/org.openhab.ui/target/org.openhab.ui-4.0.0-SNAPSHOT.jar
```

in order to try it out - please note fix applies to only `diff_last`.

If needed, revert back to official UI jar with 

```
bundle:update org.openhab.ui https://ci.openhab.org/job/openHAB-WebUI/lastSuccessfulBuild/artifact/bundles/org.openhab.ui/target/org.openhab.ui-4.0.0-SNAPSHOT.jar
```

Changes can be found [here](https://github.com/openhab/openhab-webui/compare/main...crnjan:openhab-webui:feature/diff_poc).

**Comment by cweitkamp** [comment]
> altho it's timestamp is set to starttime when returned

This is true. The additional value before expected time range will be returned with timestamp set to `starttime` and the value additional value after it with the `endtime`. But I have no answer why.

A comment in the code says:

>         // If "boundary" is true then we want to get one value before and after the requested period
>        // This is necessary for values that don't change often otherwise data will start after the start of the graph
>        // (or not at all if there's no change during the graph period)

**Comment by crnjan** [comment]
Thank you for confirming @cweitkamp! In that case I think the provided bundle above should do the job. If someone else can try it out and confirm it's giving expected results I will promote changes into a PR ...

**Comment by crnjan** [comment]
1. Should be, since `arr[1][arr[1].length - 1] - arr[1][0]` where `arr[1].length = 1` (single element returned) so we get ` arr[1][0] -  arr[1][0]` which is 0.
2. yes, was already thinking about this, we can either only take that value if it's time is exactly the same as `starttime` (where ofc there is a slim chance we actually filter out an actual datapoint if it's time would be the same) or figure out a better solution - but first would be good to get feedback with the changes so far ...

**Comment by cweitkamp** [comment]
I tested your changes and compared them against my volkszaehler.org aggregation Charts. It looks really good right now for `diff_last` mode with one exception.

Current week values are nice:
![Screenshot_20230110-122416~2](https://user-images.githubusercontent.com/5131747/211619476-f5f91d08-8df8-43a4-b186-3d11256b6ca3.jpg)
![Screenshot_20230110-123507~2](https://user-images.githubusercontent.com/5131747/211619496-9e0cfde1-8e0b-4b45-b3b7-eb7e9fc519df.jpg)

Scrolling back in time - the value for the last day (Sunday in my case) is too high - as questioned in https://github.com/openhab/openhab-webui/issues/912#issuecomment-1372539299.

![Screenshot_20230110-122425~2](https://user-images.githubusercontent.com/5131747/211619771-be670f6f-e0ba-4717-9c30-1ed0e267fb0e.jpg)
![Screenshot_20230110-123528~2](https://user-images.githubusercontent.com/5131747/211619810-790d1c77-b99b-4a11-8e99-f37e0771b3a9.jpg)

Any ideas how to solve this final issue?

**Comment by crnjan** [comment]
Good to hear it's giving expected results. Added handling for first interval, can you pls try with the updated bundle and let me know? Command to update should be the same.

**Comment by cweitkamp** [comment]
No success with your latest code changes. First bat still missing when using `diff_first` aggregation mode.

And again I am not completely happy with the results in `diff_last` mode the last value ist too high if you scroll back in time.

**Comment by crnjan** [comment]
There are no changes regarding `diff_first`, only `diff_last`. As far as I was able to test with my data it seems fine. Would you be so kind @cweitkamp and share your - chart yaml (that you use to render given data) + data for your "beginning" (so timestamp + values) so I can try to reproduce locally - thank you!

**Comment by cweitkamp** [comment]
Why "beginning"? The mistake happens because of the additional "end" value? Or did I miss something?

**Comment by crnjan** [comment]
You are right, last rendered value (calculated diff) might be off (if there are more recent values available) due 
additional `end` value. So if you render per-month data for example, i.e. for December, last "bar" rendered (for 31st of December) will take first datapoint from January to calculate diff instead of last one from December. I will update PR accordingly and let you know. Is this the issue you had in mind? Btw, have you seen https://github.com/openhab/openhab-webui/pull/1634#issuecomment-1384660817 - _rrd4j persistence service doesn't like the boundary parameter_ ?

**Comment by cweitkamp** [comment]
Yes, that is exactly what I am talking about. Thanks for the next fix.

I saw the comment and added my analysis. (ftr: I am testing with JDBC / MariaDB and have not seen the exception too.)

No, I do not expect duplicate values with the same timestamp.

**Comment by crnjan** [comment]
Did some testing and i.e.

<img width="916" alt="Screenshot 2023-01-17 at 16 15 47" src="https://user-images.githubusercontent.com/3818211/212937119-4d2b5e4d-71ec-42bb-997b-be70502f0d8c.png">

before the fix, last day of March took `end` value when doing diff calculation, which actually was stored on _April 3, 2021 9:00:04.717 PM_ - seems HP was offline at that time therefore datapoints are missing. So it took

```
      {
        "time": 1617476404717,
        "state": "23145.0"
      },
```
instead of (last datapoint of March)
```
    {
      "time": 1617091200117,
      "state": "23138.0"
    },
```
giving diff of 7 instead of 0:



<img width="903" alt="Screenshot 2023-01-17 at 16 26 22" src="https://user-images.githubusercontent.com/3818211/212939366-c31aeadd-a71e-4ad3-b04a-35a92d4f76f1.png">

@cweitkamp  - please let me know when you have time to run latest jar, thank you! And also much appreciated on taking the initiative to solve issue with `boundary` param, specially for rrd4j ...

**Comment by cweitkamp** [comment]
Working for me now. :+1: Thanks for your work.

![Screenshot_20230118-112413~2](https://user-images.githubusercontent.com/5131747/213147602-2555dcad-9172-4f21-932e-fc481d35dad9.jpg)

**Comment by benderl** [comment]
I am experiencing the same issue with current openHAB stable 3.4.4 and RRC4J persistence. As far as I know, this should be fixed in this release. Is there anything I can do to debug this?
![image](https://github.com/openhab/openhab-webui/assets/7837691/a0fb1164-8dc9-4f12-a44d-502ba6908925)

**Comment by benderl** [comment]
Oh, just realized, this fix was added to 4.0-M4. Any change to backport this into 3.4?
